### ISSUE TITLE
wrong cursor position on wrapped line when signcolumn is enabled

### ISSUE THREAD

**Issue by lacygoill** [vim/vim#8262]
**Describe the bug**

The cursor is sometimes drawn in a wrong position, on a long wrapped line when the signcolumn is enabled.

**To Reproduce**

Run this shell command:

    vim -Nu NONE -S <(cat <<'EOF'
        vim9script
        &signcolumn = 'yes'
        &linebreak = true
        &breakat = ')'
        &showbreak = '+ '
        var text: string = repeat('x', &columns - 2) .. ')abc'
        text->setline(1)
        au VimEnter * norm! $
    EOF
    )

The cursor is drawn on the `a` character.

**Expected behavior**

The cursor is drawn on the `c` character, because that's where it really is, since we've moved in that position with the normal `$` command which jumps at the end of the line, where `c` is, not `a`.  This is also confirmed by the output of the normal `ga` command:

    <c>  99,  Hex 63,  Octal 143
     ^

**Screenshots**

![gif](https://user-images.githubusercontent.com/8505073/119599521-71d6e000-bde5-11eb-8dcf-8f30fd4c81d3.gif)

**Environment**

 - Vim version: 8.2 Included patches: 1-2885
 - OS: Ubuntu 20.04.2 LTS
 - Terminal: XTerm(353)

**Additional context**

N/A

**Comment by lacygoill** [comment]
Actually, the issue is not the cursor position; it's the 2 unexpected empty cells between `)` and `abc`.

**Comment by chrisbra** [comment]
Ah, interesting problem. I think the problem is, that Vim sees the `)` and thinks it needs to reserve space for an additional `showbreak`, however the drawing logic has already added the showbreak value, so we do not need to add that space (again). Possibly fixed in https://github.com/chrisbra/vim/commit/c26b7479edfd7c76d6fc5e839e63228fef6a73d0

**Comment by lacygoill** [comment]
I can confirm that the patch fixes the issue and the test correctly fails with an unpatched binary, while it passes with a patched binary:
```diff
From c26b7479edfd7c76d6fc5e839e63228fef6a73d0 Mon Sep 17 00:00:00 2001
From: Christian Brabandt <cb@256bit.org>
Date: Wed, 26 May 2021 12:09:19 +0200
Subject: [PATCH] fix issue #8262

---
 src/drawline.c               |  4 ++++
 src/testdir/test_display.vim | 17 +++++++++++++++++
 2 files changed, 21 insertions(+)

diff --git a/src/drawline.c b/src/drawline.c
index 13be022a93d..6a19583610c 100644
--- a/src/drawline.c
+++ b/src/drawline.c
@@ -1981,6 +1981,10 @@ win_line(
 		    // TODO: is passing p for start of the line OK?
 		    n_extra = win_lbr_chartabsize(wp, line, p, (colnr_T)vcol,
 								    NULL) - 1;
+		    // We have just drawn the showbreak value,
+		    // no need to add space for it again
+		    if (vcol == vcol_sbr)
+			n_extra -= MB_CHARLEN(get_showbreak_value(wp));
 		    if (c == TAB && n_extra + col > wp->w_width)
 # ifdef FEAT_VARTABS
 			n_extra = tabstop_padding(vcol, wp->w_buffer->b_p_ts,
diff --git a/src/testdir/test_display.vim b/src/testdir/test_display.vim
index 754046f2aed..082a5b7b740 100644
--- a/src/testdir/test_display.vim
+++ b/src/testdir/test_display.vim
@@ -334,4 +334,21 @@ func Test_fold_fillchars()
   set fillchars& fdc& foldmethod& foldenable&
 endfunc
 
+func Test_display_linebreak_breakat()
+  new
+  vert resize 25
+  let _breakat=&breakat
+  setl signcolumn=yes linebreak breakat=) showbreak=+\ 
+  call setline(1, repeat('x', winwidth(0) - 2) .. ')abc')
+  " check the characters for an open fold
+  let lines = ScreenLines([1, 2], 25)
+  let expected = [
+        \ '  xxxxxxxxxxxxxxxxxxxxxxx',
+        \ '  + )abc                 '
+        \ ]
+  call assert_equal(expected, lines)
+  %bw!
+  let &breakat=_breakat
+endfunc
+
 " vim: shiftwidth=2 sts=2 expandtab
```

Thank you very much for the quick fix.

**Comment by chrisbra** [comment]
thanks for confirming. Of course the comment in the test is wrong (copy and paste too blame 🙈
